gh-125038: Crash after genexpr.gi_frame.f_locals manipulations is fixed

efimov-mikhail opened the pull request:
Some iterator checks are added for _FOR_ITER, _FOR_ITER_TIER_TWO and INSTRUMENTED_FOR_ITER bytecode implementations. TypeError is raised in case of tp_iternext == NULL. Tests on generator modifying through gi_frame.f_locals are added, both to genexpr generators and function generators.

I've added some test cases to emphasize current code behavior and save it explicitly in tests.
IMHO, there is still a room to improvement.

Previous PR becomes incorrect (https://github.com/python/cpython/pull/125051) after my typo in git commands.
I'm sorry about that.

@JelleZijlstra, Could you please review this again?
Moreover, label "needs backport to 3.13" should be added once again.

<!-- gh-issue-number: gh-125038 -->
* Issue: gh-125038
<!-- /gh-issue-number -->

efimov-mikhail committed:
gh-125038: Crash after genexpr.gi_frame.f_locals manipulations is fixed

Some iterator checks are added for _FOR_ITER,
_FOR_ITER_TIER_TWO and INSTRUMENTED_FOR_ITER bytecode implementations.
TypeError is raised in case of tp_iternext == NULL.
Tests on generator modifying through gi_frame.f_locals are added, both
to genexpr generators and function generators.

Eclips4 commented on Misc/NEWS.d/next/Core_and_Builtins/2024-10-09-13-53-50.gh-issue-125038.ffSLCz.rst:
```diff
@@ -0,0 +1,3 @@
+Fix crash when iterating over a generator expression after direct changes on ``gi_frame.f_locals``.
+Tests on generator modifying through ``gi_frame.f_locals`` are added,
+both to genexpr generators and function generators. Patch by Mikhail Efimov.
```
```suggestion
Fix crash when iterating over a generator expression after direct changes on ``gi_frame.f_locals``.
Patch by Mikhail Efimov.
```
We generally do not add notes about adding tests when fixing bugs, so we can omit that :)

efimov-mikhail replied:
Thx!

efimov-mikhail committed:
Update Misc/NEWS.d/next/Core_and_Builtins/2024-10-09-13-53-50.gh-issue-125038.ffSLCz.rst

Co-authored-by: Kirill Podoprigora <kirill.bast9@mail.ru>

JelleZijlstra commented on Python/bytecodes.c:
```diff
@@ -2806,7 +2806,16 @@ dummy_func(
+            if (iternext == NULL) {
+                _PyErr_Format(tstate, PyExc_TypeError,
+                              "'for' requires an object with "
+                              "__iter__ method, got %.100s",
```
This error is incorrect as it's looking for `__next__`, not `__iter__`. I think we can use the same error as `builtins.next()`:

```
        PyErr_Format(PyExc_TypeError, 
            "'%.200s' object is not an iterator",
            Py_TYPE(it)->tp_name);
```

efimov-mikhail committed:
gh-125038: Use TypeError with "object is not an iterator" text

efimov-mikhail committed:
gh-125038: Changes in tests on generator modifying

New tests are moved back to test_generators.py.
Tests on generator creation via FunctionType from gi_code are added.

efimov-mikhail committed:
gh-125038: CHECK_ITER instruction introduced

efimov-mikhail committed:
Merge branch 'main' into crash_fix_generators_f_locals

sobolevn commented on Lib/test/test_generators.py:
```diff
@@ -268,6 +268,84 @@ def loop():
+
+        for get_genexpr in [get_genexpr_f_locals, get_genexpr_gi_code]:
+            for seq in self.sequences:
+                err_msg_genexpr = err_msg_pattern_genexpr % type(seq).__name__
```
please use `subTest` here and in similar cases

sobolevn commented on Lib/test/test_generators.py:
```diff
@@ -268,6 +268,84 @@ def loop():
+        range(20),
+        [1, 2, 3],
+        (2,),
+        set((13, 48, 211)),
```
```suggestion
        {13, 48, 211},
```

sobolevn commented on Lib/test/test_generators.py:
```diff
@@ -268,6 +268,84 @@ def loop():
+        (2,),
+        set((13, 48, 211)),
+        frozenset((15, 8, 6)),
+        dict([(1, 2), (3, 4)]),
```
```suggestion
        {1: 2, 3: 4},
```

sobolevn commented on Lib/test/test_generators.py:
```diff
@@ -268,6 +268,84 @@ def loop():
         loop()
 
+
+class ModifyTest(unittest.TestCase):
```
question: is this a `@cpython_only` test or is this something that others should also support?

markshannon replied:
All Pythons should pass the equivalent test, but the ".0" name is a CPython implementation detail.

I'd leave this as a general test for now. Other implementations can adjust the name if needed.

efimov-mikhail committed:
gh-125038: Improvements in tests: self.subTest is used

sobolevn commented on Lib/test/test_generators.py:
```diff
@@ -268,6 +268,105 @@ def loop():
 
+
+class ModifyTest(unittest.TestCase):
+    sequences = [
```
Technically, `frozenset` and `set` are not sequences, because they don't support the `Sequence` protocol:

```python
>>> {1}[0]
<python-input-6>:1: SyntaxWarning: 'set' object is not subscriptable; perhaps you missed a comma?
  {1}[0]
Traceback (most recent call last):
  File "<python-input-6>", line 1, in <module>
    {1}[0]
    ~~~^^^
TypeError: 'set' object is not subscriptable
```

`iterables` and `non_iterables` would be better names for this test.

sobolevn commented on Lib/test/test_generators.py:
```diff
@@ -268,6 +268,105 @@ def loop():
+        def get_genexpr_gi_code(self, new_iter):
+            return self.new_gen_from_gi_code(genexpr(), new_iter)
+
+        err_msg_pattern_genexpr = "'%s' object is not an iterator"
```
I also feel like this part in test can be simplified: we can just keep the "object is not an iterator" part without the type formating, which is not essential to this test in my opinion.

But, others might feel otherwise.

sobolevn commented on Lib/test/test_generators.py:
```diff
@@ -268,6 +268,105 @@ def loop():
+        for get_genexpr in [get_genexpr_f_locals, get_genexpr_gi_code]:
+            base_name = get_genexpr.__name__
+            for seq in self.sequences:
+                test_name = f"{base_name}_seq_{type(seq).__name__}"
```
Test name can be skipped, it is enough to just pass `g_seq=g_seq` to subtests.

sobolevn commented on Lib/test/test_generators.py:
```diff
@@ -268,6 +268,105 @@ def loop():
+                err_msg = err_msg_pattern_genexpr % type(seq).__name__
+                with self.subTest(test_name, g_seq=g_seq, err_msg=err_msg):
+                    with self.assertRaisesRegex(TypeError, err_msg):
+                        list(g_seq)
```
You can also simplify this two lines into one:
```suggestion
                    self.assertRaisesRegex(TypeError, err_msg, list, g_seq)
```

sobolevn reviewed (commented):
Thanks! Just several nitpicks about tests readability from me :)

efimov-mikhail committed:
gh-125038: Further improvements in tests

sobolevn reviewed (commented):
Looks solid to me, but I am not a big compiler expert, so I won't approve :)

markshannon reviewed (approved):
Looks good now. Thanks for doing this.

JelleZijlstra commented on Include/internal/pycore_magic_number.h:
```diff
@@ -260,6 +260,7 @@ Known values:
     Python 3.14a1 3606 (Specialize CALL_KW)
     Python 3.14a1 3607 (Add pseudo instructions JUMP_IF_TRUE/FALSE)
     Python 3.14a1 3608 (Add support for slices)
+    Python 3.14a1 3609 (Add CHECK_ITER instruction)
```
@markshannon while this is fine for 3.14, the idea is to backport this change to 3.13. Are you sure it's OK to bump the magic number in a bugfix release? I looked through the 3.8 through 3.12 branches (e.g., https://github.com/python/cpython/blob/3.8/Lib/importlib/_bootstrap_external.py) and we never bumped the magic number after the beta phase. This could cause problems for people who distribute applications as pyc files.

JelleZijlstra commented on Lib/test/test_generators.py:
```diff
@@ -268,6 +268,91 @@ def loop():
         loop()
 
+
+class ModifyTest(unittest.TestCase):
```
```suggestion
class ModifyFrameLocalsTest(unittest.TestCase):
```
Name feels a bit too general

efimov-mikhail replied:
This name is not correct, since we modify generators with two different methods for now: via frame locals and by creation new generator based on the same code object.

efimov-mikhail replied:
Maybe "ModifyBaseIteratorTest"?
Or "ModifyUnderlyingIterableTest"?

JelleZijlstra commented on Python/codegen.c:
```diff
@@ -4164,6 +4164,7 @@ codegen_sync_comprehension_generator(compiler *c, location loc,
 
     if (IS_JUMP_TARGET_LABEL(start)) {
         depth++;
+        ADDOP(c, LOC(gen->iter), CHECK_ITER);
```
What if we emit `GET_ITER` here, relying on iterators obeying `iter(self) is self`? That would allow us to fix the crash without adding a new bytecode?

efimov-mikhail replied:
I think so. It will fix crash without changing magic number and bytecode values. Overhead with GET_ITER seems to be very little.

efimov-mikhail committed:
gh-125038: CHECK_ITER instruction removed, use GET_ITER instead

efimov-mikhail committed:
gh-125038: Tests are refactored

markshannon reviewed (commented):
Looks good me once you've agreed on a better name for ModifyTest.

efimov-mikhail committed:
gh-125038: More suitable name for test classes is provided

sobolevn reviewed (approved):
Thank you!
